edgelesssys/constellation
# bazel: identical containers regardless of platform

### Proposed change(s)

- bazel: identical containers regardless of platform
- automated test to find sources of non-deterministic / platform specific behavior



### Additional info

I found three sources of differences in the container images so far:

- https://github.com/bazelbuild/rules_go/issues/3581
  - TL;DR: proto rules introduce a path that differs between platforms and is embedded in any go binary that embeds the proto library
- https://github.com/bazel-contrib/rules_oci/pull/172
  - TL;DR: When using `oci_image` with a base, it's UB to also define `os` and `architecture`
- https://github.com/rmohr/bazeldnf/issues/58
  - TL;DR: bazeldnf `rpm2tar` produces non-deterministic tar files when symlinks are used
- unstripped go binaries contain debug symbols related to the platform (this seems to start with the go sdk itself)
  - TL;DR: use a version of rules_go that [supports stripping](https://github.com/bazelbuild/rules_go/pull/3527)

### How to test

As a reviewer, please comment with a list of the cli hashes you saw.
For the CLI to be built reproducible, you need to unset `build --container_prefix=FOO` in your .bazeloverwriterc.
For this, please report: The git commit you are on, if the workspace is clean (git status), the host platform (built on Ubuntu 22.04  / amd64 | built on macOS with m1), the bazel target platform (i.e. `bazel build [...] --platforms=<TARGET_PLATFORM>`), the build target with platform (`cli_enterprise_linux_amd64`).

- `bazel test //bazel/release/artifacts:artifacts_test` (Should be green, if you remove the fix for bazeldnf, it should be red)
- build the cli multiple times while specifying different platform, see if the same binary is produced:
  - `bazel build //cli:cli_enterprise_linux_amd64 && sha256sum bazel-bin/cli/cli_enterprise_linux_amd64`
  - `bazel build //cli:cli_enterprise_linux_amd64  --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64 && sha256sum bazel-bin/cli/cli_enterprise_linux_amd64`
  - `bazel build //cli:cli_enterprise_linux_amd64  --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64 && sha256sum bazel-bin/cli/cli_enterprise_linux_amd64`
  - `bazel build //cli:cli_enterprise_linux_amd64  --platforms=@io_bazel_rules_go//go/toolchain:darwin_amd64 && sha256sum bazel-bin/cli/cli_enterprise_linux_amd64`
  - `bazel build //cli:cli_enterprise_linux_amd64  --platforms=@io_bazel_rules_go//go/toolchain:darwin_arm64 && sha256sum bazel-bin/cli/cli_enterprise_linux_amd64`
  - `bazel build //bazel/release:container_sums && sha256sum bazel-bin/bazel/release/container_sums.sha256`
  - `bazel build //bazel/release:container_sums  --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64 && sha256sum bazel-bin/bazel/release/container_sums.sha256`
  - `bazel build //bazel/release:container_sums  --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64 && sha256sum bazel-bin/bazel/release/container_sums.sha256`
  - `bazel build //bazel/release:container_sums  --platforms=@io_bazel_rules_go//go/toolchain:darwin_amd64 && sha256sum bazel-bin/bazel/release/container_sums.sha256`
  - `bazel build //bazel/release:container_sums  --platforms=@io_bazel_rules_go//go/toolchain:darwin_arm64 && sha256sum bazel-bin/bazel/release/container_sums.sha256`
- build the cli on different host systems (different linux distros, different architectures, macOS) and see if the same binary is produced when building `//cli:cli_enterprise_linux_amd64` (or any of the other CLI targets that have a hardcoded platform `bazel query //cli:all`)

### Build table

| Commit                                   | Host                 | Bazel platform | Target                     | Hash                                                             |
|------------------------------------------|----------------------|----------------|----------------------------|------------------------------------------------------------------|
| 275ad5eaae0fb7f094e8e6055af9ed1682754b02 | Ubuntu 22.04 / amd64 | host (empty)   | cli_enterprise_linux_amd64 | 3881c8d15a73b77676d84851e08d85b92eebe76d8927b7c7a549812709240706 |
| 275ad5eaae0fb7f094e8e6055af9ed1682754b02 | Ubuntu 22.04 / amd64 | linux_amd64    | cli_enterprise_linux_amd64 | 3881c8d15a73b77676d84851e08d85b92eebe76d8927b7c7a549812709240706 |
| 275ad5eaae0fb7f094e8e6055af9ed1682754b02 | Ubuntu 22.04 / amd64 | linux_arm64    | cli_enterprise_linux_amd64 | 3881c8d15a73b77676d84851e08d85b92eebe76d8927b7c7a549812709240706 |
| 275ad5eaae0fb7f094e8e6055af9ed1682754b02 | Ubuntu 22.04 / amd64 | darwin_amd64   | cli_enterprise_linux_amd64 | 3881c8d15a73b77676d84851e08d85b92eebe76d8927b7c7a549812709240706 |
| 275ad5eaae0fb7f094e8e6055af9ed1682754b02 | Ubuntu 22.04 / amd64 | darwin_arm64   | cli_enterprise_linux_amd64 | 3881c8d15a73b77676d84851e08d85b92eebe76d8927b7c7a549812709240706 |
| 275ad5eaae0fb7f094e8e6055af9ed1682754b02 | Ubuntu 22.04 / amd64 | host (empty)   | container_sums             | dc695d855b6510025b0fb82fa2843adfddaaedf3f3f94052148923715774f9ec |
| 275ad5eaae0fb7f094e8e6055af9ed1682754b02 | Ubuntu 22.04 / amd64 | linux_amd64    | container_sums             | dc695d855b6510025b0fb82fa2843adfddaaedf3f3f94052148923715774f9ec |
| 275ad5eaae0fb7f094e8e6055af9ed1682754b02 | Ubuntu 22.04 / amd64 | linux_arm64    | container_sums             | dc695d855b6510025b0fb82fa2843adfddaaedf3f3f94052148923715774f9ec |
| 275ad5eaae0fb7f094e8e6055af9ed1682754b02 | Ubuntu 22.04 / amd64 | darwin_amd64   | container_sums             | dc695d855b6510025b0fb82fa2843adfddaaedf3f3f94052148923715774f9ec |
| 275ad5eaae0fb7f094e8e6055af9ed1682754b02 | Ubuntu 22.04 / amd64 | darwin_arm64   | container_sums             | dc695d855b6510025b0fb82fa2843adfddaaedf3f3f94052148923715774f9ec |


### Checklist


- [ ] Wait for fix of https://github.com/rmohr/bazeldnf/issues/58 to be merged
